### PR TITLE
[JIT] Remove Dropout during Frozon Optimization

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51589 [JIT] Remove Dropout during Frozon Optimization**

Summary: Dropout operators are only needed in training. Remove them for frozen models.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26214259](https://our.internmc.facebook.com/intern/diff/D26214259)

### CODE REVIEW

File: test/jit/test_freezing.py [hunk 1]
```diff
@@ -1517,3 +1517,36 @@ def test_optimize_freeze_module(self):
...
+                x = self.fc1(x)
+                x = nn.functional.relu(x)
+                x = self.dropout2(x)
+                x = self.fc2(x)
+                output = nn.functional.log_softmax(x, dim=1)
+                return output
+
+        mod = torch.jit.script(Net())
+        # set optimize to False here, by default freezing runs optimize_frozen_module
+        frozen_mod = torch.jit.freeze(torch.jit.script(mod.eval()), optimize=False)
```
Comment (eellison) [inline test/jit/test_freezing.py:1548]: While we're both pretty sure that it removing dropout doesn't effect results, it might be good to test it by comparing the results of the eager module with the frozen one after optimization

File: test/jit/test_freezing.py [hunk 2]
```diff
@@ -1517,3 +1517,36 @@ def test_optimize_freeze_module(self):
         # optimize_frozen_module should be run
         frozen_mod = torch.jit.freeze(torch.jit.script(mod.eval()))
         FileCheck().check_not("batch_norm").run(frozen_mod.graph)
+
+    def test_freeze_remove_dropout(self):
+        class Net(nn.Module):
+            def __init__(self):
+                super(Net, self).__init__()
+                self.conv1 = nn.Conv2d(1, 32, 3, 1)
+                self.conv2 = nn.Conv2d(32, 64, 3, 1)
```
Comment (eellison) [inline test/jit/test_freezing.py:1526]: Nit: the conv and linear here are quite big - they could be smaller without effecting the test and it would run quicker (if we run the model...). I would also consider making the module simpler / remove some layers to parse the test easier.

File: test/jit/test_freezing.py [hunk 3]
```diff
@@ -1517,3 +1517,36 @@ def test_optimize_freeze_module(self):
...
+                x = nn.functional.relu(x)
+                x = self.dropout2(x)
+                x = self.fc2(x)
+                output = nn.functional.log_softmax(x, dim=1)
+                return output
+
+        mod = torch.jit.script(Net())
+        # set optimize to False here, by default freezing runs optimize_frozen_module
+        frozen_mod = torch.jit.freeze(torch.jit.script(mod.eval()), optimize=False)
+        # inspect frozen mod
```
Comment (eellison) [inline test/jit/test_freezing.py:1549]: We should check that "aten::dropout" and "aten::feature_dropout" exist in the graph before our optimization so we know we're actually removing them

File: torch/csrc/jit/passes/remove_dropout.cpp [hunk 4]
```diff
@@ -1,3 +1,4 @@
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
```
Comment (eellison) [inline torch/csrc/jit/passes/remove_dropout.cpp:1]: I think this isn't being used

File: torch/csrc/jit/passes/frozen_graph_optimizations.cpp [hunk 5]
```diff
@@ -14,6 +15,7 @@ void OptimizeFrozenGraph(std::shared_ptr<Graph>& graph) {
     FoldFrozenConvBatchnorm(graph);
     FoldFrozenConvAddOrSub(graph);
     FoldFrozenConvMulOrDiv(graph);
+    removeDropout(graph);
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_graph_optimizations.cpp:18]: Nit: the other passes in this are run multiple times to catch different patterns of Conv-BatchNorm-Add / Conv-Add-Mul / Conv-Mul-Add. There isn't any reason to run dropout removal twice; they should be removed after the first call.

**Review by eellison — COMMENTED** [review]
Looks good! mostly just a few nits on the test

File: test/jit/test_freezing.py [hunk 6]
```diff
@@ -1517,3 +1517,25 @@ def test_optimize_freeze_module(self):
...
         frozen_mod = torch.jit.freeze(torch.jit.script(mod.eval()))
         FileCheck().check_not("batch_norm").run(frozen_mod.graph)
+
+    def test_freeze_remove_dropout(self):
+        class Net(nn.Module):
+            def __init__(self):
+                super(Net, self).__init__()
+                self.dropout = nn.Dropout(0.5)
+
+            def forward(self, x):
```
Comment (eellison) [inline test/jit/test_freezing.py:1527]: Could we add a test for aten::feature_dropout too ?

File: test/jit/test_freezing.py [hunk 7]
```diff
@@ -1517,3 +1517,25 @@ def test_optimize_freeze_module(self):
...
+                return self.dropout(x)
+
+        mod = torch.jit.script(Net())
+        script_mod = torch.jit.script(mod)
+        script_mod.eval()
+
```
Comment (eellison) [inline test/jit/test_freezing.py:1533]: have lint failure here you're going to want to fix.

**Review by eellison — APPROVED** [review]
Cool! LGTM! 

Just as an fyi - in pytorch we usually accept PRs even with comments outstanding with the assumption the non-nit comments will be addressed before landing (or there will be a comment from author on why they're not addressing it)
